### DISCUSSION TITLE
What is the best practice?

### DISCUSSION THREAD

**Discussion by l-7-l** [gftea/amqprs#78] [Q&A] [answered]
Sorry, this may be an irrelevant question. I am new to service programming. Recently, I am building a monolithic service and using rabbitmq for decoupling. I don’t know how to use amqprs in a production environment. What is the best practice. How to deal with concurrency? When should a new channel be opened? Is there any project/example recommended for production environment?

**Comment by gftea** [comment, marked as answer]
Hi,

 RabbitMQ has a good documentation https://www.rabbitmq.com/, read there for more practical advices
In general, do not use more resource than you need, 

1. Channel multiplexed on a connection,  do not use more channels than you need because it increase multiplexing, demultiplexing costs and data usage
2. do not use more queues than you need, because it increase routing costs also in the server

**Comment by michaelklishin** [comment]
Some guides I wish every new user would read early in their journey with RabbitMQ:

 * [Production Checklist](https://rabbitmq.com/production-checklist.html)
 * [Connections](https://rabbitmq.com/connections.html)
 * [Channels](https://rabbitmq.com/channels.html)
 * [Runtime](https://rabbitmq.com/runtime.html) (you don't have to tune anything in the beginning but at least be aware of what is covered in that guide)
 * [Queues](https://www.rabbitmq.com/queues.html)
 * [Quorum queues](https://www.rabbitmq.com/quorum-queues.html) (classic mirrored queues are deprecated and will be removed in RabbitMQ 4.0)
 * [Streams](https://rabbitmq.com/streams.html) and [superstreams](https://blog.rabbitmq.com/posts/2022/07/rabbitmq-3-11-feature-preview-super-streams/) (there is a [separate Rust client](https://github.com/rabbitmq/rabbitmq-stream-rust-client/) for these)
